subxt/src/backend/utils.rs

```diff
@@ -0,0 +1,291 @@
+
+pub(crate) enum PendingOrStream<T> {
+    Pending(BoxFuture<'static, Result<StreamOfResults<T>, Error>>),
+    Stream(StreamOfResults<T>),
+    Closed,
+}
+
+impl<T> std::fmt::Debug for PendingOrStream<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PendingOrStream::Pending(_) => write!(f, "Pending"),
+            PendingOrStream::Stream(_) => write!(f, "Stream"),
+            PendingOrStream::Closed => write!(f, "Closed"),
+        }
+    }
+}
+
+enum NextState<T> {
+    Stream(StreamOfResults<T>),
+    Resubscribe {
+        pending: BoxFuture<'static, Result<StreamOfResults<T>, Error>>,
+        err: Error,
+    },
+    Closed(Option<Error>),
+}
+
+impl<T> std::fmt::Debug for NextState<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Stream(_) => write!(f, "Stream"),
+            Self::Resubscribe { .. } => write!(f, "Resubscribe"),
+            Self::Closed(_) => write!(f, "Closed"),
+        }
+    }
+}
+
+/// Retry subscription.
+struct RetrySubscription<T> {
+    resubscribe: ResubscribeGetter<T>,
+    state: PendingOrStream<T>,
+}
+
+impl<T> std::marker::Unpin for RetrySubscription<T> {}
+
+impl<T> Stream for RetrySubscription<T> {
+    type Item = Result<T, Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        loop {
+            let next_state = match self.state {
+                PendingOrStream::Stream(ref mut s) => match ready!(s.poll_next_unpin(cx)) {
+                    Some(Err(err)) => {
+                        if !err.is_disconnected_will_reconnect() {
+                            NextState::Closed(Some(err))
+                        } else {
+                            NextState::Resubscribe {
+                                pending: (self.resubscribe)(),
+                                err,
+                            }
+                        }
+                    }
+                    None => NextState::Closed(None),
+                    Some(Ok(val)) => return Poll::Ready(Some(Ok(val))),
+                },
+                PendingOrStream::Pending(ref mut fut) => match ready!(fut.poll_unpin(cx)) {
+                    Ok(stream) => NextState::Stream(stream),
+                    Err(err) => {
+                        if !err.is_disconnected_will_reconnect() {
+                            NextState::Closed(Some(err))
+                        } else {
+                            NextState::Resubscribe {
+                                pending: (self.resubscribe)(),
+                                err,
+                            }
+                        }
+                    }
+                },
+                PendingOrStream::Closed => NextState::Closed(None),
+            };
+
+            match next_state {
```

Comment:
Still a shame that we need two enums; `PendingOrStream` + this `NextState` one; I was hoping we could compact it down to one enum :D

Reply:
haha, I thought it was quite readable but fine less code is probably better